picoquic/frames.c

```diff
@@ -4579,19 +4579,35 @@ uint8_t * picoquic_format_first_datagram_frame(picoquic_cnx_t* cnx, uint8_t* byt
  */
 
 typedef struct st_picoquic_datagram_buffer_argument_t {
+    picoquic_cnx_t* cnx;
     uint8_t* bytes0; /* Points to the beginning of the encoding of the datagram frame */
     uint8_t* bytes; /* Position after encoding the datagram frame type */
     uint8_t* bytes_max; /* Pointer to the end of the packet */
     uint8_t* after_data; /* Pointer to end of data written by app */
     size_t allowed_space; /* Data size from bytes to end of packet */
+    int is_active; /* Whether the application has more datagrams ready to send or not. */
+    int is_old_api; /* Whether the old buffer API was called. */
+    int was_called; /* Whether the API was called. */
 } picoquic_datagram_buffer_argument_t;
 
-uint8_t* picoquic_provide_datagram_buffer(void* context, size_t length)
+uint8_t* picoquic_provide_datagram_buffer_ex(void* context, size_t length, int is_active)
 {
     picoquic_datagram_buffer_argument_t* data_ctx = (picoquic_datagram_buffer_argument_t*)context;
     uint8_t* buffer = NULL;
 
-    if (length <= data_ctx->allowed_space) {
+    data_ctx->is_active = is_active;
+    data_ctx->was_called = 1;
+
+    if (!data_ctx->is_old_api) {
+        /* We apply the state change at this point, rather than after the return of the
+        * callback, so as to minimize "developer surprise". If the  application calls
+        * "picoquic_mark_datagram_ready" after this call, the vlaue set by the
+        * application will stick.
+        */
+        data_ctx->cnx->is_datagram_ready = is_active;
+    }
+
+    if (length <= data_ctx->allowed_space && length > 0) {
         /* Compute the length of header and length field */
```

Comment:
curious why not ``` (length > 0 && length <= data_ctx->allowed_space) ```

Reply:
Yes. "length == 0" happens in normal operation, while "length > data_ctx->allowed_space" is a programming mistake that normally never happens, so your proposed order will save some cycles. Fixed now.